doc/developer/chips/18.rst

```diff
@@ -0,0 +1,825 @@
+
+Because the chunking is on the 2nd dimension instead of the 1st, the
+corresponding NUMA localization pattern across all the array element
+indices would the following cyclic one:
+
+    =============================  ===========
+    Array Element Indices          NUMA domain
+    =============================  ===========
+    ``(0..0, 0*(2**18)..#2**18)``  0
+    ``(0..0, 1*(2**18)..#2**18)``  1
+    ``(0..0, 2*(2**18)..#2**18)``  2
+    ``(0..0, 3*(2**18)..#2**18)``  3
+    ``(1..1, 0*(2**18)..#2**18)``  0
+    ``(1..1, 1*(2**18)..#2**18)``  1
+    ``(1..1, 2*(2**18)..#2**18)``  2
+    ``(1..1, 3*(2**18)..#2**18)``  3
+    =============================  ===========
+
+A cyclic pattern is certainly achievable in the sense that the system
+can do it (or approximate it), but it is subject to the same mismatch
+issues as in the previous example: to the extent the subdomain chunking
+is different than the page-based storage localization chunking, there
+will be affinity mismatches.  In the case here that would occur if the
+page size did not evenly divide the size in bytes of ``2**18`` array
+elements.
+
+An alternative technique would instead partition across both dimensions
+at once and also localize the array storage in the same way, like this:
+
+    ====  =============================  ===========
+    Task  Array Element Indices          NUMA domain
+    ====  =============================  ===========
+    0     ``(0..0, 0*(2**19)..#2**19)``  0
+    1     ``(0..0, 1*(2**19)..#2**19)``  1
+    2     ``(1..1, 2*(2**19)..#2**19)``  2
+    3     ``(1..1, 3*(2**19)..#2**19)``  3
+    ====  =============================  ===========
+
+This is a possibility we could explore in the future if it turns out
+there are situations where it works better than the current chunking.
+But for now we are not pursuing it further.
+
+The allocate-then-localize model used for single-ddata does not
+cooperate well with network interfaces that require registered memory.
+Chapel registers the heap and other data with the NIC for ``comm=ugni``
+on Cray X\* systems and also for ``comm=gasnet, conduit=aries``.
+Registration pins virtual pages to physical pages in order to fix their
+relationship.  But changing NUMA locality necessarily means changing
+physical addresses (because NUMA is a physical characteristic), which
+requires changing the registration, which in turn means synchronizing
+with all the remote nodes which know about the registration.  In
+particular, the need for synchronization could increase the cost of
+allocation.
+
+All of the single-ddata alternatives have a lower limit on the size of
+array they can be applied to without too much waste.  Since the minimum
+unit of NUMA localization is a page, if it is to be localized an array
+needs to occupy at least as many memory pages as there are NUMA domains
+to avoid wasting space.  This can be large if NIC memory registration
+requires the use of hugepages, as it does on Cray X\* systems.
```

Comment:
To be more precise, you don't have to use hugepages for NIC registered memory on the Cray XC, but the way our comm layers are written requires the use of hugepages.